designs/full-cluster-lifecycle-api.md

```diff
@@ -0,0 +1,224 @@
+
+Right now object comparison requires a bit of transforming and parsing because our providers' code generates multiobject yaml files. In order to improve this process, we should refactor such code and start using API structs. However, this might or might not be part of the first implementation, depending on the agreed scope.
+
+### `VSphereDatacenterConfig` Validation
+
+In order to maintain some of the validations we run today in the CLI, we will need to port them to the controller. 
+
+Kubernetes offers the validation webhook for CRDs: custom code to validate objects before been accepted by the kube-api server. However, these validation are supposed to be light and fast. Most our validations require calling the vSphere API, which makes them too heavy for the validation webhook.
+
+We will implement those validations in a separate reconciler. This means that they will run after the object has already been accepted by the api server. We will store if the object has been validated or not as well as the validation error (if any) in the V`SphereDatacenterConfig` status.
+
+The cluster reconciler won’t continue with the reconciliation process until `VSphereDatacenterConfig` has been marked as validated. If not valid, it will surface the error on its own status.
+
+
+### `VSphereMachineConfig` Validation
+
+`VSphereMachineConfig` requires similar validations to the datacenter config object. We will apply the same pattern, having a separate reconciler for `VSphereMachineConfig` , the status and only reconciling the cluster once they have been validated.
+
+### New ClusterStatus 
+
+The [current Cluster Status](https://github.com/aws/eks-anywhere/blob/v0.6.0/pkg/api/v1alpha1/cluster_types.go#L287)will be modified to include extra fields and conditions. These will be used by the reconciler methods to trigger the appropriate actions and to allow users to check progress of operations.
+
+If an error occurs controllers will set `spec.ClusterStatus.FailureReason` and `spec.ClusterStatus.FailureMessage`.
+
+
+```
+type ClusterStatus struct {
+    
+    // +optional
+    FailureReason *v1alpha.ClusterStatusError `json:"failureReason,omitempty"`
+
+    // Message to help administrators understand why an operation failed 
+    // +optional
+    FailureMessage *string `json:"failureMessage,omitempty"`
+
+    // Spec has been validated
+    // +optional
+    SpecValid bool `json:"specValid,omitempty"`
+
+    // E.g. Pending, Creating, Upgrading, Deleting, Failed etc.
+    // +optional
+    Phase string `json:"phase,omitempty"`
```

Comment:
```
// +kubebuilder:validation:Enum=Pending;Updating;...
type PhaseEnum string
const (
  PhasePending PhaseEnum = "Pending'
  PhaseUpdating PhaseEnum = "Updating'
...
```

Reply:
Is this going to have a `Running` phase too?

Reply:
@TerryHowe I like the idea of the enum
However, since we still don't know which phases we will have (one of the reasons why this status is still marked as "might change"), are you ok if I add a comment saying this should be an enum and we come back and update the doc once we have the phases finalized?

Reply:
@vivek-koppuru Yeah, I think so (or `Reconciled` or whatever makes sense). But @mdsgabriel is the status's expert, so he might disagree.